File: java/src/main/java/net/snowflake/floe/FloeIv.java
package net.snowflake.floe;

import java.security.SecureRandom;

class FloeIv {
  private final byte[] bytes;

  FloeIv(byte[] bytes) {
    this.bytes = bytes;
  }

  static FloeIv generateRandom(SecureRandom random, int floeIvLength) {
    byte[] iv = new byte[floeIvLength];
    random.nextBytes(iv);
    return new FloeIv(iv);
  }

  byte[] getBytes() {
    return bytes;
  }

  int lengthInBytes() {
    return bytes.length;
  }
}

File: cpp/src/FloeIv.cpp
#include "FloeIv.hpp"
#include "floe/FloeException.hpp"
#include <openssl/rand.h>

namespace floe {

FloeIv::FloeIv(const std::vector<uint8_t>& bytes) : bytes_(bytes) {
}

FloeIv FloeIv::generateRandom(const size_t floeIvLength) {
    std::vector<uint8_t> iv(floeIvLength);
    if (RAND_bytes(iv.data(), static_cast<int>(floeIvLength)) != 1) {
        throw FloeException("Failed to generate random FLOE IV");
    }
    return FloeIv(iv);
}

const std::vector<uint8_t>& FloeIv::getBytes() const {
    return bytes_;
}

size_t FloeIv::lengthInBytes() const {
    return bytes_.size();
}

}
